### PR TITLE
add object resource info to binding

### PR DESCRIPTION
**Pull request by qianjun1993**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Add Replicas and ReplicaResourceRequirements  in binding.spec.Resource when creating

**Which issue(s) this PR fixes**:
Part of #344

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

NONE

```release-note

```

### CODE REVIEW

File: pkg/util/detector/detector.go [hunk 1]
```diff
@@ -528,15 +552,62 @@ func (d *ResourceDetector) BuildClusterResourceBinding(object *unstructured.Unst
...
+func (d *ResourceDetector) calculateResource(podSpec *corev1.PodSpec) (res util.Resource) {
+	resPtr := &res
+	for _, c := range podSpec.Containers {
+		resPtr.Add(c.Resources.Requests)
+	}
+	for _, c := range podSpec.InitContainers {
+		resPtr.Add(c.Resources.Requests)
+	}
+	return
 }
```
Comment (RainbowMango) [inline pkg/util/detector/detector.go:611]: A pod's requested resource is not simply summed, please refer to https://github.com/karmada-io/karmada/blob/30c22865aa8a0653ecfbe986d1bd78fde229d917/pkg/util/resource.go#L100-L112

File: pkg/util/detector/detector.go [hunk 2]
```diff
@@ -528,15 +552,62 @@ func (d *ResourceDetector) BuildClusterResourceBinding(object *unstructured.Unst
...
+	var podTemplateSpec *corev1.PodTemplateSpec
+	err := runtime.DefaultUnstructuredConverter.FromUnstructured(object, &podTemplateSpec)
+	if err != nil {
+		return nil, err
+	}
+	resPtr := d.calculateResource(&podTemplateSpec.Spec)
+	replicaResourceRequirements := corev1.ResourceList{
+		corev1.ResourceCPU:    *resource.NewMilliQuantity(resPtr.MilliCPU, resource.DecimalSI),
+		corev1.ResourceMemory: *resource.NewQuantity(resPtr.Memory, resource.BinarySI),
+	}
```
Comment (RainbowMango) [inline pkg/util/detector/detector.go:598]: Why not use `resPtr.ResourceList()` here?

File: pkg/util/detector/detector.go [hunk 3]
```diff
@@ -528,15 +552,62 @@ func (d *ResourceDetector) BuildClusterResourceBinding(object *unstructured.Unst
...
+				Replicas:                    replicas,
 			},
 		},
 	}
 
-	return binding
+	return binding, nil
+}
+
+// GetObjectResource get the replicas and resources of a object
```
Comment (RainbowMango) [inline pkg/util/detector/detector.go:568]:
```suggestion
// GetObjectResource get the replicas and resource requirements of a Deployment object.
```

File: pkg/util/detector/detector.go [hunk 4]
```diff
@@ -528,15 +552,62 @@ func (d *ResourceDetector) BuildClusterResourceBinding(object *unstructured.Unst
...
+func (d *ResourceDetector) GetObjectResource(object *unstructured.Unstructured) (corev1.ResourceList, int32, error) {
+	if object.GetKind() == util.DeploymentKind {
```
Comment (RainbowMango) [inline pkg/util/detector/detector.go:570]: Does this support all deployments versions, such as v1beta1/v1/v1beta2?

File: pkg/util/detector/detector.go [hunk 5]
```diff
@@ -528,15 +552,62 @@ func (d *ResourceDetector) BuildClusterResourceBinding(object *unstructured.Unst
...
 			},
 		},
 	}
 
-	return binding
+	return binding, nil
+}
+
+// GetObjectResource get the replicas and resources of a object
+func (d *ResourceDetector) GetObjectResource(object *unstructured.Unstructured) (corev1.ResourceList, int32, error) {
```
Comment (RainbowMango) [inline pkg/util/detector/detector.go:569]:
```suggestion
func (d *ResourceDetector) GetReplicaDeclaration(object *unstructured.Unstructured) (corev1.ResourceList, int32, error) {
```
How about change to this name?

**Review by RainbowMango — CHANGES_REQUESTED** [review]

File: pkg/util/detector/detector.go (hunk 5 above)
Comment (qianjun1993) [reply pkg/util/detector/detector.go:569]: ok

File: pkg/util/detector/detector.go (hunk 2 above)
Comment (qianjun1993) [reply pkg/util/detector/detector.go:598]: ok

File: pkg/util/detector/detector.go (hunk 1 above)
Comment (qianjun1993) [reply pkg/util/detector/detector.go:611]: ok

File: pkg/util/detector/detector.go (hunk 3 above)
Comment (qianjun1993) [reply pkg/util/detector/detector.go:568]: ok

File: pkg/util/detector/detector.go (hunk 4 above)
Comment (qianjun1993) [reply pkg/util/detector/detector.go:570]: I am not sure. But this judgment already exists in elsewhere。
image

**Pull request merged**